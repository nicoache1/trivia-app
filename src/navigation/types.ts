import { ParamListBase } from '@react-navigation/native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React from 'react'

import { Routes } from './routes'

export interface ParamList extends ParamListBase {
  [Routes.Welcome]: undefined
  [Routes.ConfigureGame]: undefined
  [Routes.Game]: undefined
  [Routes.Results]: undefined
  [Routes.AnimatedSplash]: undefined
}

export type SceneProps<T extends keyof ParamList> = React.FC<
  NativeStackScreenProps<ParamList, T>
>

export type NavigationProp<T extends keyof ParamList> = NativeStackScreenProps<
  ParamList,
  T
>['navigation']
