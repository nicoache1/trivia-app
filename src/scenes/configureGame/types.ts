import { Difficulties, QuestionAmounts } from '../../types/game'

export interface ConfigureGameForm {
  name: string
  amount: number
  difficulty: number
}
