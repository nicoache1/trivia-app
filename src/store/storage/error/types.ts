export interface ErrorState {
  message?: string
}

export const initialState: ErrorState = {
  message: undefined,
}
