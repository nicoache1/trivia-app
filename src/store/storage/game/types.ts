import { Difficulties, Game, QuestionAmounts } from '../../../types/game'

export interface GameState extends Game {}

export const initialState: GameState = {
  difficulty: Difficulties.Easy,
  player: 'Player',
  questionAmount: QuestionAmounts.Five,
  results: [],
}
