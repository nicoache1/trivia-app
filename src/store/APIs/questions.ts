import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import Constants from 'expo-constants'

import { deserializer, Question } from '../../types/questions'

export const baseApi = createApi({
  baseQuery: fetchBaseQuery({ baseUrl: Constants.manifest?.extra?.apiHost }),
  endpoints: builder => ({
    questions: builder.query<
      Question[],
      { amount: number; difficulty: string }
    >({
      providesTags: ['Questions'],
      query: ({ amount, difficulty }) => ({
        params: {
          amount,
          difficulty: difficulty,
          encode: 'url3986',
          type: 'boolean',
        },
        url: `api.php`,
      }),
      transformResponse: (data: any) => {
        const results = data?.results ?? []
        const parsedQuestions = results.map((item: any) => deserializer(item))
        return parsedQuestions
      },
    }),
  }),
  reducerPath: 'baseApi',
  tagTypes: ['Questions'],
})

export const {
  useLazyQuestionsQuery,
  useQuestionsQuery,
  reducer: baseApiReducer,
  reducerPath: baseApiReducerPath,
} = baseApi
