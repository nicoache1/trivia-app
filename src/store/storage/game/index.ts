import { createSlice } from '@reduxjs/toolkit'

import { restartGame, setGame, setGameResult } from './actions'
import { initialState } from './types'

const gameSlice = createSlice({
  initialState,
  name: 'game',
  reducers: {
    restartGameAction: restartGame,
    setGameAction: setGame,
    setGameResultAction: setGameResult,
  },
})

export const {
  actions: { restartGameAction, setGameAction, setGameResultAction },
  reducer: gameReducer,
} = gameSlice
