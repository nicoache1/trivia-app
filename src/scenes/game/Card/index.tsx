import React, { useCallback, useImperativeHandle } from 'react'
import { Dimensions, View } from 'react-native'
import { Gesture, GestureDetector } from 'react-native-gesture-handler'
import Animated, {
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
  withTiming,
} from 'react-native-reanimated'
import { snapPoint } from 'react-native-redash'
import { StyledText } from 'src/components/StyledText'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Question } from 'src/types/questions'

import { styles } from './styles'

interface Props {
  index: number
  item: Question
  onFinish: (option: boolean) => void
}

const { width } = Dimensions.get('window')

export const SNAP_POINTS = [-width / 1.5, 0, width / 1.5]
const Movements = {
  Center: 0,
  Left: -width * 1.5,
  Right: width * 1.5,
}

export const Card = React.forwardRef<
  { triggerAnimation: (option: boolean) => void },
  Props
>(({ index, item, onFinish }, ref) => {
  const {
    Theme: { colors },
  } = useTheme()

  const translationX = useSharedValue(0)

  const triggerAnimation = useCallback(
    (option: boolean) => {
      translationX.value = withTiming(
        option ? Movements.Left : Movements.Right,
        {},
        () => {
          runOnJS(onFinish)(option)
        },
      )
    },
    [onFinish, translationX],
  )

  useImperativeHandle(
    ref,
    () => ({
      triggerAnimation,
    }),
    [triggerAnimation],
  )

  const gesture = Gesture.Pan()
    .onUpdate(e => {
      translationX.value = e.translationX
    })
    .onEnd(({ velocityX }) => {
      const nearestPoint = snapPoint(translationX.value, velocityX, SNAP_POINTS)
      let toValue =
        nearestPoint === SNAP_POINTS[0]
          ? Movements.Left
          : nearestPoint === SNAP_POINTS[2]
          ? Movements.Right
          : Movements.Center
      translationX.value = withSpring(toValue, {})
      if (toValue !== Movements.Center) {
        runOnJS(onFinish)(Movements.Left ? true : false)
      }
    })
    .withTestId('PanGesture')

  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: translationX.value }],
    }
  })

  return (
    <GestureDetector gesture={gesture}>
      <Animated.View
        style={[
          styles.container,
          { backgroundColor: colors.WHITE, zIndex: index },
          animatedStyle,
        ]}>
        <View style={styles.categoryContainer}>
          <View
            style={[
              styles.categoryLabel,
              {
                backgroundColor: colors.SECONDARY,
              },
            ]}>
            <StyledText
              testID="CardCategory"
              style={styles.text}
              typography={AppTypography.P1_BOLD}
              color={ColorPalette.WHITE}
              allowFontScaling>
              {item.category}
            </StyledText>
          </View>
        </View>
        <View style={styles.questionContainer}>
          <StyledText
            testID="CardQuestion"
            style={styles.text}
            typography={AppTypography.H2}>
            {item.question}
          </StyledText>
        </View>
      </Animated.View>
    </GestureDetector>
  )
})
