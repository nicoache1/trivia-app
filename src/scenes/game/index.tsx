import CheckImage from 'assets/check.png'
import CrossImage from 'assets/cross.png'
import React, { useEffect, useLayoutEffect, useRef, useState } from 'react'
import { View } from 'react-native'
import { useSelector } from 'react-redux'
import { Background } from 'src/components/Background'
import { SceneContainer } from 'src/components/SceneContainer'
import { StyledContainer } from 'src/components/StyledContainer'
import { StyledText } from 'src/components/StyledText'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useAppDispatch } from 'src/store'
import { useLazyQuestionsQuery } from 'src/store/APIs/questions'
import { setGameResultAction } from 'src/store/storage/game'
import { selectGameConfiguration } from 'src/store/storage/game/selectors'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Question } from 'src/types/questions'
import { v4 as uuidv4 } from 'uuid'

import { Card } from './Card'
import { FinishGame } from './FinishGame'
import { GameButton } from './GameButton'
import { styles } from './styles'

export const Game: SceneProps<Routes.Game> = ({ navigation }) => {
  const game = useSelector(selectGameConfiguration)
  const {
    Theme: { colors },
  } = useTheme()
  const dispatch = useAppDispatch()
  const [triggerQuery, { isLoading, isFetching }] = useLazyQuestionsQuery()

  const ref = useRef<{
    triggerAnimation: (option: boolean) => void
  }>(null)

  const [cards, setCards] = useState<Question[]>([])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  useEffect(() => {
    async function getQuestions() {
      const data = await triggerQuery({
        amount: game.questionAmount,
        difficulty: game.difficulty,
      }).unwrap()
      setCards(data)
    }

    getQuestions()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const onFinish = (option: boolean) => {
    const question = cards[cards.length - 1]

    dispatch(
      setGameResultAction({
        results: [
          {
            correct: question.correctAnswer === option,
            id: uuidv4(),
            question,
          },
        ],
      }),
    )

    setCards(prevState => {
      return prevState.slice(0, prevState.length - 1)
    })
  }

  const onPressButton = (option: boolean) => () => {
    ref.current?.triggerAnimation(option)
  }

  const finishGame = cards.length === 0 && !isLoading && !isFetching

  const finishCallback = () => navigation.navigate(Routes.Results)

  return (
    <>
      {finishGame && <FinishGame callback={finishCallback} />}
      <SceneContainer
        barStyle={'light-content'}
        style={[styles.container, { backgroundColor: colors.SECONDARY }]}>
        <Background />
        <StyledContainer style={styles.playerContainer}>
          <StyledText color={ColorPalette.WHITE} typography={AppTypography.H1}>
            {game.player}
          </StyledText>
        </StyledContainer>
        <View style={styles.cardContainer}>
          {cards.map((item, index) => (
            <Card
              ref={index === cards.length - 1 ? ref : undefined}
              key={item.id}
              item={item}
              index={index}
              onFinish={onFinish}
            />
          ))}
        </View>
        <StyledContainer style={styles.answerOptionContainer}>
          <GameButton imageSource={CheckImage} onPress={onPressButton(true)} />
          <GameButton imageSource={CrossImage} onPress={onPressButton(false)} />
        </StyledContainer>
      </SceneContainer>
    </>
  )
}
