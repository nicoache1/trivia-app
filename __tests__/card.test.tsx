import { render } from '@testing-library/react-native'
import React from 'react'
import { State } from 'react-native-gesture-handler'
import {
  fireGestureHandler,
  getByGestureTestId,
} from 'react-native-gesture-handler/jest-utils'
import * as ReactNativeRedash from 'react-native-redash'
import { Difficulties } from 'src/types/game'
import { Question } from 'src/types/questions'

import { Card, SNAP_POINTS } from '../src/scenes/game/Card/index'

const timeout = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

jest.mock('react-native-redash')

const mockFn = jest.fn()
const mockIndex = 0
const mockQuestion: Question = {
  category: 'Category',
  correctAnswer: false,
  difficulty: Difficulties.Easy,
  id: 'mock-id',
  incorrectAnswers: [true],
  question: 'Question',
  type: true,
}

describe('Test swipe gestures', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('Expect callback function to have been called after swipe gesture to left', async () => {
    const snapPointMock = jest.spyOn(ReactNativeRedash, 'snapPoint')
    snapPointMock.mockReturnValue(SNAP_POINTS[0])

    render(<Card index={mockIndex} item={mockQuestion} onFinish={mockFn} />)

    const gesture = getByGestureTestId('PanGesture')

    fireGestureHandler(gesture, [
      { state: State.END, translationX: 300, velocityX: 120 },
    ])

    await timeout(3000)
    expect(mockFn).toBeCalled()
  })

  test('Expect callback function to have been called after swipe gesture to right', async () => {
    const snapPointMock = jest.spyOn(ReactNativeRedash, 'snapPoint')
    snapPointMock.mockReturnValue(SNAP_POINTS[2])

    render(<Card index={mockIndex} item={mockQuestion} onFinish={mockFn} />)

    const gesture = getByGestureTestId('PanGesture')

    fireGestureHandler(gesture, [
      { state: State.END, translationX: 300, velocityX: 120 },
    ])

    await timeout(3000)
    expect(mockFn).toBeCalled()
  })

  test('Expect callback function to not have been called', async () => {
    const snapPointMock = jest.spyOn(ReactNativeRedash, 'snapPoint')
    snapPointMock.mockReturnValue(SNAP_POINTS[1])

    render(<Card index={mockIndex} item={mockQuestion} onFinish={mockFn} />)

    const gesture = getByGestureTestId('PanGesture')

    fireGestureHandler(gesture, [
      { state: State.END, translationX: 300, velocityX: 120 },
    ])

    await timeout(3000)
    expect(mockFn).toBeCalledTimes(0)
  })
})

test('Expect card to be rendered correctly', async () => {
  const renderedSnapshot = render(
    <Card index={mockIndex} item={mockQuestion} onFinish={mockFn} />,
  ).toJSON()

  expect(renderedSnapshot).toMatchSnapshot()
})
