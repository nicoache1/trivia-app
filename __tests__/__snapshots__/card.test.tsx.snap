// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect card to be rendered correctly 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      },
    }
  }
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 16,
      "height": 667,
      "overflow": "hidden",
      "padding": 20,
      "position": "absolute",
      "transform": Array [
        Object {
          "translateX": 0,
        },
      ],
      "width": "100%",
      "zIndex": 0,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 16,
            "height": 40,
            "justifyContent": "center",
            "minWidth": 120,
          },
          Object {
            "backgroundColor": "rgb(108,92,231)",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        color="WHITE"
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
        testID="CardCategory"
        typography="P1_BOLD"
      >
        Category
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 2,
      }
    }
  >
    <Text
      accessible={true}
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontStyle": "normal",
          "fontWeight": "700",
          "textAlign": "center",
        }
      }
      testID="CardQuestion"
      typography="H2"
    >
      Question
    </Text>
  </View>
</View>
`;
